core/executor/src/wasm_executor.rs

```diff
@@ -19,100 +19,243 @@
 
 use wasmi::{
 	Module, ModuleInstance, MemoryInstance, MemoryRef, TableRef, ImportsBuilder, ModuleRef,
 	memory_units::Pages, RuntimeValue::{I32, I64, self},
 };
-use state_machine::{Externalities, ChildStorageKey};
+use state_machine::Externalities;
 use crate::error::{Error, Result};
-use codec::Encode;
+use codec::{Encode, Decode};
 use primitives::{
 	blake2_128, blake2_256, twox_64, twox_128, twox_256, ed25519, sr25519, Pair, crypto::KeyTypeId,
-	offchain, hexdisplay::HexDisplay, sandbox as sandbox_primitives, H256, Blake2Hasher,
+	offchain, hexdisplay::HexDisplay, sandbox as sandbox_primitives, Blake2Hasher,
 };
 use trie::{TrieConfiguration, trie_types::Layout};
 use crate::sandbox;
 use crate::allocator;
 use log::trace;
+use wasm_interface::{FunctionContext, HostFunctions, Pointer, WordSize, Sandbox, MemoryId};
 
 #[cfg(feature="wasm-extern-trace")]
 macro_rules! debug_trace {
 	( $( $x:tt )* ) => ( trace!( $( $x )* ) )
 }
+
 #[cfg(not(feature="wasm-extern-trace"))]
 macro_rules! debug_trace {
 	( $( $x:tt )* ) => ()
 }
 
-struct FunctionExecutor<'e, E: Externalities<Blake2Hasher> + 'e> {
+struct FunctionExecutor {
 	sandbox_store: sandbox::Store,
 	heap: allocator::FreeingBumpHeapAllocator,
 	memory: MemoryRef,
 	table: Option<TableRef>,
-	ext: &'e mut E,
-	hash_lookup: HashMap<Vec<u8>, Vec<u8>>,
 }
 
-impl<'e, E: Externalities<Blake2Hasher>> FunctionExecutor<'e, E> {
-	fn new(m: MemoryRef, heap_base: u32, t: Option<TableRef>, e: &'e mut E) -> Result<Self> {
+impl FunctionExecutor {
+	fn new(m: MemoryRef, heap_base: u32, t: Option<TableRef>) -> Result<Self> {
 		Ok(FunctionExecutor {
 			sandbox_store: sandbox::Store::new(),
 			heap: allocator::FreeingBumpHeapAllocator::new(m.clone(), heap_base),
 			memory: m,
 			table: t,
-			ext: e,
-			hash_lookup: HashMap::new(),
 		})
 	}
 }
 
-impl<'e, E: Externalities<Blake2Hasher>> sandbox::SandboxCapabilities for FunctionExecutor<'e, E> {
+impl sandbox::SandboxCapabilities for FunctionExecutor {
 	fn store(&self) -> &sandbox::Store {
 		&self.sandbox_store
 	}
 	fn store_mut(&mut self) -> &mut sandbox::Store {
 		&mut self.sandbox_store
 	}
-	fn allocate(&mut self, len: u32) -> Result<u32> {
+	fn allocate(&mut self, len: WordSize) -> Result<Pointer<u8>> {
 		self.heap.allocate(len)
 	}
-	fn deallocate(&mut self, ptr: u32) -> Result<()> {
+	fn deallocate(&mut self, ptr: Pointer<u8>) -> Result<()> {
 		self.heap.deallocate(ptr)
 	}
-	fn write_memory(&mut self, ptr: u32, data: &[u8]) -> Result<()> {
-		self.memory.set(ptr, data).map_err(Into::into)
+	fn write_memory(&mut self, ptr: Pointer<u8>, data: &[u8]) -> Result<()> {
+		self.memory.set(ptr.into(), data).map_err(Into::into)
+	}
+	fn read_memory(&self, ptr: Pointer<u8>, len: WordSize) -> Result<Vec<u8>> {
+		self.memory.get(ptr.into(), len as usize).map_err(Into::into)
+	}
+}
+
+impl FunctionContext for FunctionExecutor {
+	fn read_memory_into(&self, address: Pointer<u8>, dest: &mut [u8]) -> std::result::Result<(), String> {
```

Comment:
Nit: There are way more `std::result::Result<_, _>` than just `Result`. I'd prefer to properly alias one or both results.